genesis/engine/solvers/rigid/gjk_decomp.py

```diff
@@ -0,0 +1,3394 @@
+
+    @ti.func
+    def func_add_edge_to_horizon(self, i_b, i_f, i_e):
+        horizon_nedges = self.polytope[i_b].horizon_nedges
+        self.polytope_horizon_data[i_b, horizon_nedges].edge_idx = i_e
+        self.polytope_horizon_data[i_b, horizon_nedges].face_idx = i_f
+        self.polytope[i_b].horizon_nedges += 1
+
+        flag = 0
+        return flag
+
+    @ti.func
+    def func_get_edge_idx(self, i_b, i_f, i_v):
+        """
+        Get the edge index from the face, starting from the vertex i_v.
+        If the face is comprised of [v1, v2, v3], the edges are:
+        [v1, v2], [v2, v3], [v3, v1].
+
+        Therefore, if i_v is v1, the edge index is 0,
+        and if i_v is v2, the edge index is 1,
+        and if i_v is v3, the edge index is 2.
+        """
+        verts = self.polytope_faces[i_b, i_f].verts_idx
+        ret = 2
+        if verts[0] == i_v:
+            ret = 0
+        elif verts[1] == i_v:
+            ret = 1
+        return ret
+
+    @ti.func
+    def func_delete_face_from_polytope(self, i_b, i_f):
+        """
+        Delete the face from the polytope.
+        """
+        face_map_idx = self.polytope_faces[i_b, i_f].map_idx
+        if face_map_idx >= 0:
+            last_face_idx = self.polytope_faces_map[i_b][self.polytope[i_b].nfaces_map - 1]
+            # Make the map to point to the last face
+            self.polytope_faces_map[i_b][face_map_idx] = last_face_idx
+            # Change map index of the last face
+            self.polytope_faces[i_b, last_face_idx].map_idx = face_map_idx
+
+            # Decrease the number of faces in the polytope
+            self.polytope[i_b].nfaces_map -= 1
+
+        # Mark the face as deleted
+        self.polytope_faces[i_b, i_f].map_idx = -2
+
+    @ti.func
+    def func_epa_insert_vertex_to_polytope(self, i_b, obj1_point, obj2_point, obj1_id, obj2_id, minkowski_point):
+        """
+        Copy vertex information into the polytope.
+        """
+        n = self.polytope[i_b].nverts
+        self.polytope_verts[i_b, n].obj1 = obj1_point
+        self.polytope_verts[i_b, n].obj2 = obj2_point
+        self.polytope_verts[i_b, n].id1 = obj1_id
+        self.polytope_verts[i_b, n].id2 = obj2_id
+        self.polytope_verts[i_b, n].mink = minkowski_point
+        self.polytope[i_b].nverts += 1
+        return n
+
+    @ti.func
+    def func_epa_init_polytope_2d(self, i_ga, i_gb, i_b):
+        """
+        Create the polytope for EPA from a 1-simplex (line segment).
+        Return 0 when successful.
+        """
+        flag = 0
+
+        # Get the simplex vertices
+        v1 = self.gjk_simplex_vertex[i_b, 0].mink
+        v2 = self.gjk_simplex_vertex[i_b, 1].mink
+        diff = v2 - v1
+
+        # Find the element in [diff] with the smallest magnitude,
+        # because it will give us the largest cross product below
+        min_val = ti.abs(diff[0])
+        min_i = 0
+        for i in range(1, 3):
+            if ti.abs(diff[i]) < min_val:
```

Comment:
add `abs_diff_i` or something to avoid computing it twice. Even if I agree it should not change anything overall. Just a good practice.